File: benchmark-scala/src/main/scala/ArithmeticTest.scala
import scala.util.Sorting
import scala.collection.mutable.ListBuffer
import scala.compat.Platform

/**
 * ArithmeticTest$.
 * 
 * @author anavarro - 6 Sep 2009
 * 
 */
object ArithmeticTest
{
    val NbTests = 5
    val NbOfExculsionMinMax = 2
    val NbArithmeticTests = 1
    val IntMax = 1 * 1000 * 1000 * 1000
    val DoubleMin = 10 * 1000 * 1000 * 1000.0
    val DoubleMax = 11 * 1000 * 1000 * 1000.0
    val LongMin = 10 * 1000 * 1000 * 1000L
    val LongMax = 11 * 1000 * 1000 * 1000L
                                                            
    
    def averageTimeWithoutMinMax(executionTimes: Seq[Long], numberOfMinMaxToRemove: Int): Long = 
    {   
        var average: Long = 0
        val minMax = if (numberOfMinMaxToRemove > 0) numberOfMinMaxToRemove else 0
        val orderedExecutionTimes = if (minMax != 0) Sorting.stableSort(executionTimes) else executionTimes
        for (i <- minMax.until(orderedExecutionTimes.size - minMax))
        {
           average += orderedExecutionTimes(i)
        }
        average / (orderedExecutionTimes.size - 2 * minMax)
    }


    def relativeDeviationTimeWithoutMinMax(executionTimes: Seq[Long], numberOfMinMaxToRemove: Int): Double = 
    {
        val avgTime: Long = averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove)
        var deviation: Double = 0
        val minMax = if (numberOfMinMaxToRemove > 0) numberOfMinMaxToRemove else 0
        val orderedExecutionTimes = if (minMax != 0) Sorting.stableSort(executionTimes) else executionTimes
        for (i <- minMax.until(orderedExecutionTimes.size - minMax))
        {
          deviation += Math.pow(orderedExecutionTimes(i) - avgTime, 2)
        }
        (((Math.sqrt(deviation / (orderedExecutionTimes.size - 2 * minMax)) / avgTime * 100) * 100).asInstanceOf[Int]) / 100d
    }

    
    def testIntArithmetic: Long = 
    {
        val start = Platform.currentTime
        var intResult = 1
        var i = 1
        while (i < IntMax)
        {
          intResult -= i
          i += 1
          intResult += i
          i += 1
          intResult *= i
          i += 1
          intResult /= i
          i += 1
        }
        val end = Platform.currentTime
        val executionTime = end - start
        println("[ArithmeticTest], Calculation of " + NbArithmeticTests  + " var intResult = 1  var i = 1  while (i < " + IntMax + ") {intResult -= i  i += 1  intResult += i i += 1  intResult *= i  i += 1  intResult /= i  i += 1},intResult=" + intResult + ", snapshot time," + executionTime)
        executionTime
    }

    def testDoubleArithmetic: Long = 
    {
        val start = Platform.currentTime
        var doubleResult: Double = DoubleMin
        var i : Double = DoubleMin
        while (i < DoubleMax)
        {
          doubleResult -= i
          i += 1
          doubleResult += i
          i += 1
          doubleResult *= i
          i += 1
          doubleResult /= i
          i += 1
        }
        val end = Platform.currentTime
        val executionTime = end - start
        println("[ArithmeticTest], Calculation of " + NbArithmeticTests  + " var doubleResult : Double = doubleMin  var i : Double = doubleMin  while (i < " + DoubleMax + ") {doubleResult -= i  i += 1  doubleResult += i i += 1  doubleResult *= i  i += 1  doubleResult /= i  i += 1}, doubleResult=" + doubleResult + ", snapshot time," + executionTime)
        executionTime
    }
        
    def testLongArithmetic: Long = 
    {
        val start = Platform.currentTime
        var longResult: Long = LongMin
        var i : Long = LongMin
        while (i < LongMax)
        {
          longResult -= i
          i += 1
          longResult += i
          i += 1
          longResult *= i
          i += 1
          longResult /= i
          i += 1
        }
        val end = Platform.currentTime
        val executionTime = end - start
        println("[ArithmeticTest], Calculation of " + NbArithmeticTests  + " var longResult : Long = longMin  var i = 1  while (i < " + LongMax + ") {longResult -= i  i += 1  longResult += i i += 1  longResult *= i  i += 1  longResult /= i  i += 1}, longResult=" + longResult + ", snapshot time," + executionTime)
        executionTime
    }

    
    /**
     * main.
     * 
     * @param args
     */
    def main(args: Array[String]) =
    {
        val nbTests = if (args.size >= 1) args(0).toInt else NbTests;
        val nbOfExclusionMinMax = if (args.size >= 2) args(1).toInt else NbOfExculsionMinMax
        
        val executionTimes = new ListBuffer[Long]
        executionTimes.clear
        for (i <- 1 to nbTests)
        {
            executionTimes + testIntArithmetic 
        }
        println("[ArithmeticTest], Calculation of " + NbArithmeticTests + "  var intResult = 1  var i = 1  while (i < " + IntMax + ") {intResult -= i++  intResult += i++  intResult *= i++  intResult /= i++ },, average time," + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes.foldLeft(Long.MinValue)(_ min _) + ", max time," + executionTimes.foldLeft(Long.MinValue)(_ max _) + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));   
        
        executionTimes.clear
        for (i <- 1 to nbTests)
        {
            executionTimes + testDoubleArithmetic
        }
        println("[ArithmeticTest], Calculation of " + NbArithmeticTests  + " var doubleResult : Double = doubleMin  var i : Double = doubleMin  while (i < " + DoubleMax + ") {doubleResult -= i  i += 1  doubleResult += i i += 1  doubleResult *= i  i += 1  doubleResult /= i  i += 1},, average time," + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes.foldLeft(Long.MinValue)(_ min _) + ", max time," + executionTimes.foldLeft(Long.MinValue)(_ min _) + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));   
        
        executionTimes.clear
        for (i <- 1 to nbTests)
        {
            executionTimes + testLongArithmetic
        }
        println("[ArithmeticTest], Calculation of " + NbArithmeticTests  + " var longResult : Long = longMin  var i = 1  while (i < " + LongMax + ") {longResult -= i  i += 1  longResult += i i += 1  longResult *= i  i += 1  longResult /= i  i += 1},, average time," + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + executionTimes.foldLeft(Long.MinValue)(_ min _) + ", max time," + executionTimes.foldLeft(Long.MinValue)(_ max _) + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));   
        
    }
}

File: benchmark-java/src/main/java/ArithmeticTest.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * ArithmeticTest.
 * 
 * @author anavarro122404 - 25 sept. 07
 * 
 * 
 * <!-- $Id: ArithmeticTest.java,v 1.1 2007/10/15 17:58:12 anavarro Exp $ -->.
 * 
 */
public final class ArithmeticTest
{
    /**
     * NB_TESTS
     */
    private static final int    NB_TESTS                = 5;
    
    /**
     * NB_OF_EXCLUSION_MIN_MAX
     */
    private static final int    NB_OF_EXCLUSION_MIN_MAX = 2;
    
    /**
     * NB_CREATION_TESTS
     */
    private static final int    NB_ARITHMETIC_TESTS     = 1;
    
    /**
     * INT_MAX
     */
    private static final int    INT_MAX                 = 1 * 1000 * 1000 * 1000;
    
    /**
     * DOUBLE_MIN
     */
    private static final double DOUBLE_MIN              = 10 * 1000 * 1000 * 1000.0;
    
    /**
     * DOUBLE_MAX
     */
    private static final double DOUBLE_MAX              = 11 * 1000 * 1000 * 1000.0;
    
    /**
     * LONG_MIN
     */
    private static final long   LONG_MIN                = 10 * 1000 * 1000 * 1000L;
    
    /**
     * LONG_MAX
     */
    private static final long   LONG_MAX                = 11 * 1000 * 1000 * 1000L;
    
    /**
     * Constructor.
     * 
     */
    private ArithmeticTest()
    {
        //
    }
    
    
    /**
     * averageTimeWithoutMinMax.
     * 
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static long averageTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        long average = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            average += executionTimes.get(i);
        }
        average = average / (executionTimes.size() - minMax * 2);
        return average;
    }
    
    
    /**
     * relativeDeviationTimeWithoutMinMax.
     *
     * @param executionTimes
     * @param numberOfMinMaxToRemove
     * @return
     */
    @SuppressWarnings("boxing")
    public static double relativeDeviationTimeWithoutMinMax(final List<Long> executionTimes, final int numberOfMinMaxToRemove)
    {
        final long averageTimeWithoutMinMax = averageTimeWithoutMinMax(executionTimes, numberOfMinMaxToRemove);
        long deviation = 0;
        final int minMax = (numberOfMinMaxToRemove > 0) ? numberOfMinMaxToRemove : 0; 
        if (minMax != 0)
        {
            Collections.sort(executionTimes);
        }
        for (int i = minMax; i < executionTimes.size() - minMax; i++)
        {
            deviation += Math.pow(executionTimes.get(i) - averageTimeWithoutMinMax, 2);
        }
        return ((int) ((Math.sqrt(deviation / (executionTimes.size() - 2 * minMax)) / averageTimeWithoutMinMax * 100) * 100)) / 100d;
    }
    

    
    /**
     * testIntArithmetic.
     * 
     * @return
     */
    public static long testIntArithmetic()
    {
        long start = System.currentTimeMillis();
        start = System.currentTimeMillis();
        int intResult = 1;
        int i = 1;
        while (i < INT_MAX)
        {
            intResult -= i++;
            intResult += i++;
            intResult *= i++;
            intResult /= i++;
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[ArithmeticTest], Calculation of " + NB_ARITHMETIC_TESTS + " int intResult = 1  int i = 1  while (i < " + INT_MAX
                + ") {intResult -= i++  intResult += i++  intResult *= i++  intResult /= i++},intResult=" + intResult + ", snapshot time,"
                + executionTime);
        return executionTime;
    }
    
    /**
     * testDoubleArithmetic.
     * 
     * @return
     */
    public static long testDoubleArithmetic()
    {
        long start = System.currentTimeMillis();
        start = System.currentTimeMillis();
        double doubleResult = DOUBLE_MIN;
        double i = DOUBLE_MIN;
        while (i < DOUBLE_MAX)
        {
            doubleResult -= i++;
            doubleResult += i++;
            doubleResult *= i++;
            doubleResult /= i++;
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[ArithmeticTest], Calculation of " + NB_ARITHMETIC_TESTS + " double doubleResult = " + DOUBLE_MIN + " double i = "
                + DOUBLE_MIN + "  while (i < " + DOUBLE_MAX
                + ") {doubleResult -= i++  doubleResult += i++  doubleResult *= i++  doubleResult /= i++ },doubleResult=" + doubleResult
                + ", snapshot time," + executionTime);
        return executionTime;
    }
    
    /**
     * testLongArithmetic.
     * 
     * @return
     */
    public static long testLongArithmetic()
    {
        long start = System.currentTimeMillis();
        start = System.currentTimeMillis();
        long longResult = LONG_MIN;
        long i = LONG_MIN;
        while (i < LONG_MAX)
        {
            longResult -= i++;
            longResult += i++;
            longResult *= i++;
            longResult /= i++;
        }
        long end = System.currentTimeMillis();
        long executionTime = end - start;
        System.out.println("[ArithmeticTest], Calculation of " + NB_ARITHMETIC_TESTS + " long longResult = " + LONG_MIN + "  long i = " + LONG_MIN
                + "  while (i < " + LONG_MAX + ") { longResult -= i++  longResult += i++  longResult *= i++  longResult /= i++ },longResult="
                + longResult + ", snapshot time," + executionTime);
        return executionTime;
    }
    
    
    
    /**
     * main.
     * 
     * @param args
     */
    @SuppressWarnings("boxing")
    public static void main(final String[] args)
    {
        final int nbTests = (args != null && args.length >= 1 && args[0] != null) ?  Integer.valueOf(args[0]): NB_TESTS;
        final int nbOfExclusionMinMax = (args != null && args.length >= 2 && args[1] != null) ? Integer.valueOf(args[1]) : NB_OF_EXCLUSION_MIN_MAX;
        
        final List<Long> executionTimes = new ArrayList<Long>(nbTests);
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(ArithmeticTest.testIntArithmetic());
        System.out.println("[ArithmeticTest], Calculation of " + NB_ARITHMETIC_TESTS + " int intResult = 1  int i = 1  while (i < " + INT_MAX
                + ") {intResult -= i++  intResult += i++  intResult *= i++  intResult /= i++ },, average time," + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax)
                + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes) + ", relative deviation time,"
                + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(ArithmeticTest.testDoubleArithmetic());
        System.out.println("[ArithmeticTest], Calculation of " + NB_ARITHMETIC_TESTS + " double doubleResult = " + DOUBLE_MIN + "  double i = "
                + DOUBLE_MIN + "  while (i < " + DOUBLE_MAX
                + ") {doubleResult -= i++  doubleResult += i++  doubleResult *= i++  doubleResult /= i++ },, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
        for (int i = nbTests; i != 0; i--)
            executionTimes.add(ArithmeticTest.testLongArithmetic());
        System.out.println("[ArithmeticTest], Calculation of " + NB_ARITHMETIC_TESTS + " long longResult = " + LONG_MIN + "  long i = " + LONG_MIN
                + "  while (i < " + LONG_MAX + ") { longResult -= i++  longResult += i++  longResult *= i++  longResult /= i++ },, average time,"
                + averageTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax) + ", min time," + Collections.min(executionTimes) + ", max time," + Collections.max(executionTimes)
                + ", relative deviation time," + relativeDeviationTimeWithoutMinMax(executionTimes, nbOfExclusionMinMax));
        executionTimes.clear();
        
    }
}
